app/src/main/java/com/gaurav/avnc/ui/vnc/VncActivity.kt

```diff
@@ -321,12 +321,13 @@ class VncActivity : AppCompatActivity() {
         //need Mouse right-click events. It is hardcoded to act as back-press, without
         //giving apps a chance to handle it. For better or worse, they set the 'source'
         //for such key events to Mouse, enabling the following workarounds.
+        val device = InputDevice.getDevice(keyEvent.deviceId)
         if (Build.VERSION.SDK_INT >= 23 &&
             keyEvent.keyCode == KeyEvent.KEYCODE_BACK &&
-            keyEvent.source == InputDevice.SOURCE_MOUSE) {
+            device.sources or InputDevice.SOURCE_MOUSE == device.sources) {
```

Comment:
We can simplify the `if` condition to:
```kotlin
    if (keyEvent.keyCode == KeyEvent.KEYCODE_BACK &&
        InputDevice.getDevice(keyEvent.deviceId).supportsSource(InputDevice.SOURCE_MOUSE)) {
```

1. [supportsSource()](https://developer.android.com/reference/android/view/InputDevice?hl=en#supportsSource(int)) can be used for source test.
2. API check can be dropped. Originally, I was emulating `ACTION_BUTTON_PRESS` & `ACTION_BUTTON_RELEASE` here,
   which are only available on API 23+.

Curiously, does this source test work for you (as the source in key event was Keyboard)?

Reply:
Updated the condition.

Here is the sources values of some devices, both keyboard cover touchpad and bluetooth mouse support `SOURCE_MOUSE`:
```
*Keyboard Cover Touchpad: 12290
Keyboard Cover: 257
*Bluetooth Mouse Back (Thumb) Button: 8451
Bluetooth Keyboard: 16778001
Android Back: 769
One Hand Operator+ Back: 769
```